llvm/lib/CheerpUtils/SIMDLowering.cpp

```diff
@@ -0,0 +1,284 @@
+	bool add = (id == Intrinsic::vector_reduce_add || id == Intrinsic::vector_reduce_fadd) ? true : false;
+	// Only accept normal, not floats, for now.
+	assert(id == Intrinsic::vector_reduce_add || id == Intrinsic::vector_reduce_mul);
+	Value* total = values[0];
+	for (int i = 1; i < amount; i++)
+	{
+		if (add)
+			total = Builder.CreateAdd(total, values[i]);
+		else
+			total = Builder.CreateMul(total, values[i]);
+	}
+	I.replaceAllUsesWith(total);
+	deleteList.push_back(&I);
+	return false;
+}
+
+bool SIMDLoweringPass::lowerBitShift(Instruction& I)
+{
+	// This function will lower bit shift operations that take a vector as their second operand,
+	// since WebAssembly does not support this.
+	Value* secondOp = I.getOperand(1);
+	if (!secondOp->getType()->isVectorTy())
+		return false;
+	Intrinsic::ID intrID;
+	unsigned opcode = I.getOpcode();
+	if (opcode == Instruction::Shl)
+		intrID = Intrinsic::cheerp_wasm_shl;
+	else if (opcode == Instruction::AShr)
+		intrID = Intrinsic::cheerp_wasm_shr_s;
+	else
+		intrID = Intrinsic::cheerp_wasm_shr_u;
+	assert(isa<ConstantDataVector>(secondOp) || isa<Instruction>(secondOp));
+	IRBuilder<> Builder(&I);
+	const ConstantDataVector* cdv = dyn_cast<ConstantDataVector>(secondOp);
+	const CallInst* callInst = dyn_cast<CallInst>(secondOp);
+	if (cdv && cdv->isSplat())
+	{
+		ConstantInt* splatValue = cast<ConstantInt>(cdv->getSplatValue());
+		ConstantInt* shiftValue = Builder.getInt32(splatValue->getZExtValue());
+		std::vector<Type *> argTypes = { I.getType(), I.getType(), shiftValue->getType() };
+		Function* intrinsic = Intrinsic::getDeclaration(I.getModule(), intrID, argTypes);
+		CallInst* ci = Builder.CreateCall(intrinsic, {I.getOperand(0), shiftValue});
+		I.replaceAllUsesWith(ci);
+		deleteList.push_back(&I);
+	}
+	else if (callInst && callInst->getIntrinsicID() == Intrinsic::cheerp_wasm_splat)
+	{
+		std::vector<Type *> argTypes = { I.getType(), I.getType(), callInst->getOperand(0)->getType() };
+		Function* intrinsic = Intrinsic::getDeclaration(I.getModule(), intrID, argTypes);
+		CallInst* ci = Builder.CreateCall(intrinsic, {I.getOperand(0), callInst->getOperand(0) });
+		I.replaceAllUsesWith(ci);
+		deleteList.push_back(&I);
+	}
+	else
+		// General lower case.
+		return lowerGeneralUnsupportedVectorOperation(I);
+	return false;
+}
+
+bool SIMDLoweringPass::lowerSplat(Instruction &I)
+{
+	// Try to see if this instruction was originally a splat.
+	// We know this if we find a shufflevector instruction, that has a zero mask, and it's first operand
+	// is an insert element used only once, and which has an undef/poison operand.
+	const ShuffleVectorInst& svi = cast<ShuffleVectorInst>(I);
+	if (svi.isZeroEltSplat())
+	{
+		Value* firstOp = svi.getOperand(0);
+		if (!firstOp->hasOneUse())
+			return false;
+		if (const InsertElementInst* iei = dyn_cast<InsertElementInst>(firstOp))
+		{
+			if (isa<UndefValue>(iei->getOperand(0)))
```

Comment:
This condition I don't think it's actually needed, since we are fine with any kind of iei->getOperand(0)

(It's by convention UndefValue in the current IR, but that might/will change, and we don't rely on that)

Reply:
I changed this to instead check that the 3rd operand is a 0 value, that's the relevant part.